File: src/main/java/dev/fabby/com/commands/BaseCommand.java
package dev.fabby.com.commands;

import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;

public abstract class BaseCommand {

    public abstract String name();

    public abstract String description();
    public abstract String[] aliases();

    public abstract String neededPermission();

    public abstract int minArgs();

    public abstract String usage();

    public abstract boolean consoleAllowed();

    public abstract void execute(final Player p, final String[] args);

    public abstract void executeAsConsole(final CommandSender sender, final String[] args);

    public abstract boolean customTabCompleter();

    public abstract TabCompleter tabCompleter();
}

File: src/main/java/dev/fabby/com/commands/BaseCommand.kt
package dev.fabby.com.commands

import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player

abstract class BaseCommand {
    abstract fun name(): String
    abstract fun description(): String
    abstract fun aliases(): Array<String>
    abstract fun neededPermission(): String
    abstract fun minArgs(): Int
    abstract fun usage(): String
    abstract fun consoleAllowed(): Boolean
    abstract fun execute(p: Player, args: Array<String>)
    abstract fun executeAsConsole(sender: CommandSender, args: Array<String>)
    abstract fun customTabCompleter(): Boolean
    abstract fun tabCompleter(): TabCompleter?
}
